aptos-move/framework/aptos-framework/sources/overloadable_fungible_asset.move

```diff
@@ -0,0 +1,169 @@
+            string::utf8(b"overloadable_fungible_asset"),
+            string::utf8(b"dispatchable_deposit"),
+        );
+        // Verify that caller type matches callee type so wrongly typed function cannot be registered.
+        assert!(function_info::check_dispatch_type_compatibility(
+            &dispatcher_deposit_function_info,
+            &deposit_function
+        ), error::invalid_argument(EDEPOSIT_FUNCTION_SIGNATURE_MISMATCH));
+
+        assert!(!object::can_generate_delete_ref(constructor_ref), error::invalid_argument(EOBJECT_IS_DELETABLE));
+        assert!(!exists<OverloadFunctionStore>(object::address_from_constructor_ref(constructor_ref)), error::already_exists(EALREADY_REGISTERED));
+
+        // Freeze the FungibleStore to force usign the new overloaded api.
+        let extend_ref = object::generate_extend_ref(constructor_ref);
+        fungible_asset::set_global_frozen_flag(&extend_ref, true);
+
+        let store_obj = &object::generate_signer(constructor_ref);
+
+        // Store the overload function hook.
+        move_to<OverloadFunctionStore>(store_obj, OverloadFunctionStore {
+            withdraw_function,
+		    deposit_function,
+            transfer_ref: fungible_asset::generate_transfer_ref(constructor_ref),
+        });
+    }
+
+    /// Withdraw `amount` of the fungible asset from `store` by the owner.
+    ///
+    /// The semantics of deposit will be governed by the function specified in OverloadFunctionStore.
+    public fun withdraw<T: key>(
+        owner: &signer,
+        store: Object<T>,
+        amount: u64,
+    ): FungibleAsset acquires OverloadFunctionStore {
+        let metadata_addr = object::object_address(&fungible_asset::store_metadata(store));
+        let owner_address = signer::address_of(owner);
+        assert!(exists<OverloadFunctionStore>(metadata_addr), error::not_found(EFUNCTION_STORE_NOT_FOUND));
+        let overloadable_store = borrow_global<OverloadFunctionStore>(metadata_addr);
+        dispatchable_withdraw(
+            owner_address,
+            store,
+            amount,
+            &overloadable_store.transfer_ref,
+            &overloadable_store.withdraw_function,
+        )
+    }
+
+    /// Deposit `amount` of the fungible asset to `store`.
+    ///
+    /// The semantics of deposit will be governed by the function specified in OverloadFunctionStore.
+    public fun deposit<T: key>(
+        store: Object<T>,
+        fa: FungibleAsset
+    ) acquires OverloadFunctionStore {
+        let metadata_addr = object::object_address(&fungible_asset::store_metadata(store));
+        assert!(exists<OverloadFunctionStore>(metadata_addr), error::not_found(EFUNCTION_STORE_NOT_FOUND));
+        let overloadable_store = borrow_global<OverloadFunctionStore>(metadata_addr);
+        dispatchable_deposit(
+            store,
+            fa,
+            &overloadable_store.transfer_ref,
+            &overloadable_store.deposit_function,
+        );
+    }
```

Comment:
That's an interesting case. To avoid double assess maybe we should just use the transfer api so at most one side pays the tax?

Reply:
What is the issue? I already pay two taxes in the US ... income and sales :sob:

Reply:
@davidiw my main concern here is ensuring that a transfer can have either a known amount sent or received

else the accounting for DeFi breaks, since you don't have a known amount deposited into a pool (known receive amount) or withdrawn from a pool (known send amount)